### PR TITLE
[WIP] Study possibility of moving from js-yaml to yaml

### PR DESCRIPTION
**Pull request by belgattitude**
**What's the problem this PR addresses?**

When running `yarn set version` or `yarn plugin import`, the comments of the `.yarnrc.yml` are lost. 

After finding a previous issue https://github.com/yarnpkg/berry/issues/1463 and reading @paul-soporan comments, I was curious about possibilities.

His idea about the [enhanced-yaml](https://github.com/paul-soporan/enhanced-yaml) seems convincing to me, I keep this P/R open not sure I'd work on it though

...

**How did you fix it?**

Not yet, I'm just evaluating step by steps. Changing the yaml parser affects installs (the `yarn.lock` is read) so I'm not even sure I'd like to open a pandora box. 

So here's some steps:

- [x] Try a parser that supports comments, ie [yaml](https://www.npmjs.com/package/yaml).
- [x] Spot differences between js-yaml and yaml. 
- [x] Estimate the size overhead. (+900kb, ? to check)
- [x] Bench the reading the yarn.lock (much slower)
- [ ] Gather feedbacks (sharing here https://github.com/eemeli/yaml/discussions/358#discussioncomment-2212132, let's see maybe I'm not using it like it should)
- [ ] Ultimately preserve comments :)

**Early findings**

- The binary jumped from ` 2.06Mb` to `2.1Mb`. 
- The speed is an issue (not for .yarnrc.yml, more that the parser will ideally have to deal with the yarn.lock too)

**Checklist**

- [ ] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [ ] I have set the packages that need to be released for my changes to be effective.

- [ ] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-parsers/package.json [hunk 1]
```diff
@@ -4,11 +4,11 @@
...
   "main": "./sources/index.ts",
   "dependencies": {
-    "js-yaml": "^3.10.0",
-    "tslib": "^1.13.0"
+    "tslib": "^1.13.0",
+    "yaml": "^2.0.0-10"
   },
   "devDependencies": {
-    "@types/js-yaml": "^3.11.1",
+    "js-yaml": "^3.10.0",
```
Comment (belgattitude) [inline packages/yarnpkg-parsers/package.json:11]: TODO: clear this, just for comparing the parsers on the yarn.lock

File: packages/yarnpkg-parsers/sources/syml.ts [hunk 2]
```diff
@@ -141,10 +140,22 @@ function parseViaJsYaml(source: string) {
   if (LEGACY_REGEXP.test(source))
     return parseViaPeg(source);
 
-  const value = safeLoad(source, {
-    schema: FAILSAFE_SCHEMA,
-    json: true,
-  });
+  // js-yaml@3 doesn't parse `key: true` as a boolean, but as a string.
+  // as far as I tested it makes the lock file invalid after install.
+  const reviver = (key: string, value: unknown) => {
```
Comment (belgattitude) [inline packages/yarnpkg-parsers/sources/syml.ts:145]: It's more a difference... The yaml parser parse the native value, js-yaml keep a string. Without this it broke the lock file, but I guess a refactor can be done in the stringify too, thus keeping what yaml does.

**Comment by arcanis** [comment]
Note that @paul-soporan already looked into it and it seemed quite difficult to achieve without duplicating dependencies or slowing down installs:

https://github.com/yarnpkg/berry/issues/1463#issuecomment-687302655

**Comment by merceyz** [comment]
Making this a draft due to the "WIP" in the title

File: packages/yarnpkg-parsers/tests/syml.test.js [hunk 3]
```diff
@@ -21,13 +25,48 @@ describe(`Syml parser`, () => {
...
+    const yamlParsed = parseWithYaml(lock, reviver, {
+      uniqueKeys: false,
+      schema: 'failsafe',
+      customTags: []
+    })
+    const endYaml = performance.now();
+    const yamlTime = endYaml - startYaml;
+
+    expect(yamlParsed).toStrictEqual(jsYamlParsed);
+    expect(yamlTime).toBeLessThan(jsYamlTime);
```
Comment (belgattitude) [inline packages/yarnpkg-parsers/tests/syml.test.js:70]: Current finding:

Expected:  < 64.09353400021791 ms (js-yaml)
Received:   513.0383080020547 ms (yaml)

Yes that might be a real blocker (at least for parsing yarn.lock which is not required). Adding both libs might not be a good move :(

**Comment by belgattitude** [comment]
Mmm, I'll have a look with fresh head and maybe take another path. If it's just the yarnrc, it might possible to hack the comments while still using js-yaml... Let's see.

**Comment by belgattitude** [comment]
I don't feel there's a good solution right now.  So I'll close.

**Pull request closed without merging**